gobblin-modules/gobblin-kafka-08/src/main/java/org/apache/gobblin/metrics/kafka/KafkaKeyValueProducerPusher.java

```diff
@@ -80,17 +98,48 @@ public KafkaKeyValueProducerPusher(String brokers, String topic) {
    */
   public void pushMessages(List<Pair<K, V>> messages) {
     for (Pair<K, V> message: messages) {
-      this.producer.send(new ProducerRecord<>(topic, message.getKey(), message.getValue()), (recordMetadata, e) -> {
+      this.futures.offer(this.producer.send(new ProducerRecord<>(topic, message.getKey(), message.getValue()), (recordMetadata, e) -> {
         if (e != null) {
           log.error("Failed to send message to topic {} due to exception: ", topic, e);
         }
-      });
+      }));
+    }
+
+    //Once the low watermark of numFuturesToBuffer is hit, start flushing messages from the futures
+    // buffer. In order to avoid blocking on newest messages added to futures queue, we only invoke future.get() on
+    // the oldest messages in the futures buffer. The number of messages to flush is same as the number of messages added
+    // in the current call. Note this does not completely avoid calling future.get() on the newer messages e.g. when
+    // multiple threads enter the if{} block concurrently, and invoke flush().
+    if (this.futures.size() >= this.numFuturesToBuffer) {
+      flush(messages.size());
+    }
+  }
+
+  /**
+   * Flush any records that may be present in the producer buffer upto a maximum of <code>numRecordsToFlush</code>.
+   * This method is needed since Kafka 0.8 producer does not have a flush() API. In the absence of the flush()
+   * implementation, records which are present in the buffer but not in-flight may not be delivered at all when close()
+   * is called, leading to data loss.
+   * @param numRecordsToFlush
+   */
+  private void flush(long numRecordsToFlush) {
+    log.info("Flushing records from producer buffer");
```

Comment:
Can you make this log and the one at the end debug level or print it only when `numRecordsToFlush == Long.MAX_VALUE`? I think this may flood the log since flush() is called whenever the futures queue has hit the limit.